File: chart-generator/generate_charts.py
#!/usr/bin/env python3
"""
HdrHistogram Chart Generator for Benchmark Results
Generates latency percentile comparison charts from wrk benchmark data
"""

import sys
import os
import re
from pathlib import Path
import matplotlib.pyplot as plt
import matplotlib
matplotlib.use('Agg')  # Non-interactive backend

# Configuration
CONFIGS = ['6by1', '3by2', '2by3', '2by2']
CONFIG_COLORS = {
    '6by1': '#FF6B6B',  # Red
    '3by2': '#4ECDC4',  # Teal
    '2by3': '#45B7D1',  # Blue
    '2by2': '#FFA07A'   # Orange
}
CONFIG_LABELS = {
    '6by1': '6 pods × 1 CPU',
    '3by2': '3 pods × 2 CPU',
    '2by3': '2 pods × 3 CPU',
    '2by2': '2 pods × 2 CPU'
}


def parse_wrk_output(file_path, min_percentile=95.0):
    """Parse wrk output file and extract latency percentile data
    
    Args:
        file_path: Path to the wrk output file
        min_percentile: Minimum percentile to include (default: 95.0 for P95+)
    """
    percentiles = []
    latencies = []
    
    with open(file_path, 'r') as f:
        content = f.read()
    
    # Find the detailed percentile section
    in_section = False
    for line in content.split('\n'):
        if 'Detailed Percentile spectrum:' in line:
            in_section = True
            continue
        
        if in_section:
            # Stop at the summary section
            if line.strip().startswith('#[Mean'):
                break
            
            # Parse data lines: "     0.144     0.000000            1         1.00"
            parts = line.strip().split()
            if len(parts) == 4:
                try:
                    latency = float(parts[0])  # milliseconds
                    percentile = float(parts[1]) * 100  # convert to percentage
                    
                    # Only include data at or above the minimum percentile
                    if percentile >= min_percentile:
                        latencies.append(latency)
                        percentiles.append(percentile)
                except ValueError:
                    continue
    
    return percentiles, latencies


def generate_comparison_chart(results_dir, test_type, output_file):
    """Generate a comparison chart for all configurations"""
    plt.figure(figsize=(14, 8))
    
    # Determine title, endpoint, and percentile range based on test type
    if test_type == 'simple':
        title = 'Extreme Tail Latency (P99.9 to P99.9999) - Simple JSON Endpoint'
        endpoint = '/json'
        min_percentile = 99.9  # Focus on P99.9+ for simple endpoint (the "nines")
        xlim_min = 99.9
    else:
        title = 'Extreme Tail Latency (P99.9 to P99.9999) - CPU-Intensive Endpoint'
        endpoint = '/waitWithPrimeFactor'
        min_percentile = 99.9  # Focus on P99.9+ for CPU-intensive endpoint (the "nines")
        xlim_min = 99.9
    
    found_data = False
    
    # Plot each configuration
    for config in CONFIGS:
        file_path = results_dir / f"{config}_{test_type}.txt"
        
        if not file_path.exists():
            print(f"  ⚠️  Skipping {config}: file not found")
            continue
        
        percentiles, latencies = parse_wrk_output(file_path, min_percentile)
        
        if not percentiles or not latencies:
            print(f"  ⚠️  Skipping {config}: no data found")
            continue
        
        # Plot the data
        plt.plot(percentiles, latencies, 
                label=CONFIG_LABELS[config],
                color=CONFIG_COLORS[config],
                linewidth=2,
                marker='',
                alpha=0.8)
        
        print(f"  ✅ Plotted {config}: {len(percentiles)} data points")
        found_data = True
    
    if not found_data:
        print(f"  ❌ No data found for {test_type} endpoint")
        plt.close()
        return False
    
    # Customize the plot
    plt.xlabel('Percentile (%)', fontsize=12, fontweight='bold')
    plt.ylabel('Latency (milliseconds)', fontsize=12, fontweight='bold')
    plt.title(title, fontsize=14, fontweight='bold', pad=20)
    plt.grid(True, alpha=0.3, linestyle='--')
    plt.legend(loc='upper left', fontsize=11, framealpha=0.9)
    
    # Focus on tail latencies with appropriate range for test type
    plt.xlim(xlim_min, 100)
    
    # Add custom ticks to highlight the "nines" (P99.9, P99.99, P99.999, P99.9999)
    custom_ticks = [99.9, 99.99, 99.999, 99.9999, 100.0]
    plt.xticks(custom_ticks, ['99.9%', '99.99%', '99.999%', '99.9999%', '100%'])
    
    # Add subtitle with endpoint info
    plt.text(0.5, -0.12, f'Endpoint: {endpoint}',
             ha='center', va='top',
             transform=plt.gca().transAxes,
             fontsize=10, style='italic', color='gray')
    
    # Tight layout
    plt.tight_layout()
    
    # Save the chart
    plt.savefig(output_file, dpi=150, bbox_inches='tight')
    plt.close()
    
    print(f"  📸 Chart saved: {output_file}")
    return True


def extract_summary_stats(file_path):
    """Extract summary statistics from wrk output"""
    stats = {}
    
    with open(file_path, 'r') as f:
        content = f.read()
    
    # Extract key metrics
    patterns = {
        'requests_sec': r'Requests/sec:\s+([\d.]+)',
        'latency_avg': r'Latency\s+([^\s]+)',
        'latency_max': r'Latency.*Max\s+([^\s]+)',
        'p50': r'50\.000%\s+([\d.]+)',
        'p75': r'75\.000%\s+([\d.]+)',
        'p90': r'90\.000%\s+([\d.]+)',
        'p99': r'99\.000%\s+([\d.]+)',
        'p999': r'99\.900%\s+([\d.]+)',
    }
    
    for key, pattern in patterns.items():
        match = re.search(pattern, content)
        if match:
            stats[key] = match.group(1)
    
    return stats


def generate_summary_table(results_dir, output_file):
    """Generate a summary comparison table"""
    with open(output_file, 'w') as f:
        f.write("# Benchmark Summary\n\n")
        
        for test_type in ['simple', 'cpu']:
            endpoint = '/json' if test_type == 'simple' else '/waitWithPrimeFactor'
            f.write(f"\n## {endpoint}\n\n")
            f.write("| Config | Req/sec | Avg Latency | P50 | P90 | P99 | P99.9 |\n")
            f.write("|--------|---------|-------------|-----|-----|-----|-------|\n")
            
            for config in CONFIGS:
                file_path = results_dir / f"{config}_{test_type}.txt"
                
                if not file_path.exists():
                    f.write(f"| {CONFIG_LABELS[config]} | N/A | N/A | N/A | N/A | N/A | N/A |\n")
                    continue
                
                stats = extract_summary_stats(file_path)
                f.write(f"| {CONFIG_LABELS[config]} | "
                       f"{stats.get('requests_sec', 'N/A')} | "
                       f"{stats.get('latency_avg', 'N/A')} | "
                       f"{stats.get('p50', 'N/A')}ms | "
                       f"{stats.get('p90', 'N/A')}ms | "
                       f"{stats.get('p99', 'N/A')}ms | "
                       f"{stats.get('p999', 'N/A')}ms |\n")
        
        f.write("\n---\n")
        f.write(f"\n*Generated: {Path(results_dir).name}*\n")


def main():
    """Main function"""
    print("=" * 50)
    print("HdrHistogram Chart Generator (Python)")
    print("=" * 50)
    print()
    
    # Get results directory from command line
    if len(sys.argv) < 2:
        print("❌ Error: No results directory specified")
        print("Usage: python3 generate_charts.py <results-directory>")
        sys.exit(1)
    
    results_dir = Path(sys.argv[1])
    
    if not results_dir.exists():
        print(f"❌ Error: Results directory not found: {results_dir}")
        sys.exit(1)
    
    print(f"📁 Results directory: {results_dir}")
    
    # Create output directory
    output_dir = results_dir / 'charts'
    output_dir.mkdir(exist_ok=True)
    print(f"📁 Output directory: {output_dir}")
    print()
    
    # Generate charts for each test type
    success_count = 0
    
    for test_type in ['simple', 'cpu']:
        endpoint = 'Simple JSON' if test_type == 'simple' else 'CPU-Intensive'
        output_file = output_dir / f"comparison_{test_type}.png"
        
        print(f"📈 Generating chart for: {endpoint} endpoint")
        
        if generate_comparison_chart(results_dir, test_type, output_file):
            success_count += 1
        
        print()
    
    # Generate summary table
    summary_file = output_dir / 'summary.md'
    print(f"📊 Generating summary table: {summary_file}")
    generate_summary_table(results_dir, summary_file)
    print(f"  ✅ Summary saved")
    print()
    
    # Final summary
    print("=" * 50)
    if success_count > 0:
        print(f"✅ Chart generation complete!")
        print()
        print(f"📁 Charts saved in: {output_dir}")
        print()
        print(f"Generated files:")
        if (output_dir / 'comparison_simple.png').exists():
            print(f"  - comparison_simple.png (Simple JSON endpoint)")
        if (output_dir / 'comparison_cpu.png').exists():
            print(f"  - comparison_cpu.png (CPU-intensive endpoint)")
        print(f"  - summary.md (Summary statistics)")
        print()
        print(f"View charts: open {output_dir}/")
    else:
        print("❌ No charts were generated")
        sys.exit(1)
    print("=" * 50)


if __name__ == '__main__':
    main()

File: chart-generator/generate-charts.js
#!/usr/bin/env node

import { chromium } from 'playwright';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Configuration
const HDR_HISTOGRAM_URL = 'http://hdrhistogram.github.io/HdrHistogram/plotFiles.html';
const RESULTS_DIR = process.argv[2] || 'benchmark-results';
const OUTPUT_DIR = path.join(RESULTS_DIR, 'charts');

// Colors for different configurations
const CONFIG_COLORS = {
  '6by1': '#FF6B6B',  // Red
  '3by2': '#4ECDC4',  // Teal
  '2by3': '#45B7D1',  // Blue
  '2by2': '#FFA07A'   // Orange
};

console.log('====================================');
console.log('HdrHistogram Chart Generator');
console.log('====================================\n');

if (!fs.existsSync(RESULTS_DIR)) {
  console.error(`❌ Error: Results directory not found: ${RESULTS_DIR}`);
  console.error('Usage: node generate-charts.js <results-directory>');
  process.exit(1);
}

// Create output directory
if (!fs.existsSync(OUTPUT_DIR)) {
  fs.mkdirSync(OUTPUT_DIR, { recursive: true });
}

console.log(`📁 Results directory: ${RESULTS_DIR}`);
console.log(`📁 Output directory: ${OUTPUT_DIR}\n`);

// Convert wrk output to HdrHistogram format
function parseWrkOutput(filePath) {
  const content = fs.readFileSync(filePath, 'utf-8');
  const lines = content.split('\n');
  
  const histogramData = [];
  let inDetailedSection = false;
  
  for (const line of lines) {
    if (line.includes('Detailed Percentile spectrum:')) {
      inDetailedSection = true;
      continue;
    }
    
    if (inDetailedSection) {
      // Parse lines like: "  0.500    0.125  50.00%"
      const match = line.trim().match(/^([\d.]+)\s+([\d.]+)\s+([\d.]+)%/);
      if (match) {
        const percentile = parseFloat(match[1]);
        const latencyMs = parseFloat(match[2]) * 1000; // Convert to microseconds
        histogramData.push({ percentile, latency: latencyMs });
      }
      
      if (line.includes('Mean')) {
        break;
      }
    }
  }
  
  // If no detailed data, extract from summary
  if (histogramData.length === 0) {
    const latencyLine = lines.find(l => l.includes('Latency') && l.includes('avg'));
    if (latencyLine) {
      // Simple extraction for basic visualization
      console.log(`  ⚠️  Using summary data for ${path.basename(filePath)}`);
    }
  }
  
  return histogramData;
}

// Generate comparison chart data
function generateComparisonData(configs, testType) {
  const comparisonData = {};
  
  for (const config of configs) {
    const filePath = path.join(RESULTS_DIR, `${config}_${testType}.txt`);
    if (fs.existsSync(filePath)) {
      const data = parseWrkOutput(filePath);
      if (data.length > 0) {
        comparisonData[config] = data;
      }
    }
  }
  
  return comparisonData;
}

// Main function to generate charts
async function generateCharts() {
  console.log('🌐 Launching browser...\n');
  
  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext({
    viewport: { width: 1920, height: 1080 }
  });
  
  const page = await context.newPage();
  
  // Navigate to HdrHistogram plotter
  console.log('📊 Loading HdrHistogram plotter...\n');
  await page.goto(HDR_HISTOGRAM_URL, { waitUntil: 'networkidle' });
  
  const configs = ['6by1', '3by2', '2by3', '2by2'];
  const testTypes = ['simple', 'cpu'];
  
  for (const testType of testTypes) {
    console.log(`\n📈 Generating chart for: ${testType === 'simple' ? 'Simple JSON' : 'CPU-Intensive'} endpoint\n`);
    
    // Reload page for fresh chart
    await page.goto(HDR_HISTOGRAM_URL, { waitUntil: 'networkidle' });
    await page.waitForTimeout(2000);
    
    let fileIndex = 0;
    
    for (const config of configs) {
      const filePath = path.join(RESULTS_DIR, `${config}_${testType}.txt`);
      
      if (!fs.existsSync(filePath)) {
        console.log(`  ⚠️  Skipping ${config}: file not found`);
        continue;
      }
      
      console.log(`  📂 Processing ${config}...`);
      
      try {
        // Read the benchmark file
        const fileContent = fs.readFileSync(filePath, 'utf-8');
        
        // Find the file input element
        const fileInput = await page.locator('input[type="file"]').first();
        
        // Create a temporary file with proper histogram format
        const tempFile = path.join(OUTPUT_DIR, `temp_${config}_${testType}.hgrm`);
        
        // Convert wrk output to HdrHistogram format
        // For simplicity, we'll upload the raw file and let the tool parse it
        // Or we can create a proper .hgrm file
        const hgrm = convertToHgrm(fileContent, config);
        fs.writeFileSync(tempFile, hgrm);
        
        // Upload file
        await fileInput.setInputFiles(tempFile);
        await page.waitForTimeout(1500);
        
        // Clean up temp file
        fs.unlinkSync(tempFile);
        
        console.log(`  ✅ Uploaded ${config}`);
        fileIndex++;
        
      } catch (error) {
        console.log(`  ❌ Error processing ${config}: ${error.message}`);
      }
    }
    
    // Wait for chart to render
    await page.waitForTimeout(3000);
    
    // Take screenshot
    const screenshotPath = path.join(OUTPUT_DIR, `comparison_${testType}.png`);
    await page.screenshot({ 
      path: screenshotPath,
      fullPage: true 
    });
    
    console.log(`\n  📸 Chart saved: ${screenshotPath}\n`);
  }
  
  await browser.close();
  
  console.log('\n====================================');
  console.log('✅ Chart generation complete!');
  console.log('====================================\n');
  console.log(`📁 Charts saved in: ${OUTPUT_DIR}\n`);
  console.log('Generated charts:');
  console.log('  - comparison_simple.png (Simple JSON endpoint)');
  console.log('  - comparison_cpu.png (CPU-intensive endpoint)\n');
}

// Convert wrk output to HdrHistogram format
function convertToHgrm(wrkOutput, configName) {
  const lines = wrkOutput.split('\n');
  let hgrm = `#[Histogram for ${configName}]\n`;
  hgrm += `#[StartTime: ${Date.now()} (seconds since epoch)]\n`;
  
  // Find latency distribution
  let inDistribution = false;
  
  for (const line of lines) {
    if (line.includes('Latency Distribution')) {
      inDistribution = true;
      continue;
    }
    
    if (inDistribution && line.trim()) {
      // Parse lines like: "50%    1.23ms"
      const match = line.trim().match(/([\d.]+)%\s+([\d.]+)(ms|us|s)/);
      if (match) {
        const percentile = parseFloat(match[1]);
        let latency = parseFloat(match[2]);
        const unit = match[3];
        
        // Convert to microseconds
        if (unit === 'ms') latency *= 1000;
        else if (unit === 's') latency *= 1000000;
        
        // HdrHistogram format: Value, Percentile, TotalCount, 1/(1-Percentile)
        const totalCount = Math.floor(percentile * 100);
        hgrm += `${latency.toFixed(3)} ${(percentile/100).toFixed(6)} ${totalCount} ${(1/(1-percentile/100)).toFixed(2)}\n`;
      }
    }
    
    // Also capture detailed percentile spectrum if available
    if (line.includes('Detailed Percentile spectrum:')) {
      break;
    }
  }
  
  return hgrm;
}

// Run the generator
generateCharts().catch(error => {
  console.error('❌ Error:', error);
  process.exit(1);
});
